File: app/src/main/java/com/druger/aboutwork/utils/recycler/EndlessRecyclerViewScrollListener.java
package com.druger.aboutwork.utils.recycler;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

/**
 * Created by druger on 04.01.2018.
 */

public abstract class EndlessRecyclerViewScrollListener extends RecyclerView.OnScrollListener {

    private int visibleThreshold = 2;
    private int currentPage = 0;
    private int previousTotalItemCount = 0;
    private boolean loading = true;
    private int startingPageIndex = 0;

    private RecyclerView.LayoutManager layoutManager;

    public EndlessRecyclerViewScrollListener(LinearLayoutManager layoutManager) {
        this.layoutManager = layoutManager;
    }

    @Override
    public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
        int lastVisibleItemPosition = 0;
        int totalItemCount = layoutManager.getItemCount();

        if (layoutManager instanceof  LinearLayoutManager) {
            lastVisibleItemPosition = ((LinearLayoutManager) layoutManager).findLastVisibleItemPosition();
        }

        if (totalItemCount < previousTotalItemCount) {
            currentPage = startingPageIndex;
            previousTotalItemCount = totalItemCount;
            if (totalItemCount == 0) {
                loading = true;
            }
        }

        if (loading && (totalItemCount > previousTotalItemCount)) {
            loading = false;
            previousTotalItemCount = totalItemCount;
        }

        if (!loading
                && (lastVisibleItemPosition + visibleThreshold) >= totalItemCount
                && dy > 0) {
            currentPage++;
            onLoadMore(currentPage);
            loading = true;
        }
    }

    public abstract void onLoadMore(int page);

    private void resetPageCount(int page) {
        previousTotalItemCount = 0;
        loading = true;
        currentPage = page;
        onLoadMore(currentPage);
    }

    public void resetPageCount() {
        resetPageCount(0);
    }
}

File: app/src/main/java/com/druger/aboutwork/utils/recycler/EndlessRecyclerViewScrollListener.kt
package com.druger.aboutwork.utils.recycler

import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

/**
 * Created by druger on 04.01.2018.
 */

abstract class EndlessRecyclerViewScrollListener(private val layoutManager: LinearLayoutManager) :
    RecyclerView.OnScrollListener() {

    private val visibleThreshold = 5
    private var currentPage = 0
    private var previousTotalItemCount = 0
    private var loading = true
    private val startingPageIndex = 0
    private var pages = 0

    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
        super.onScrolled(recyclerView, dx, dy)
        val lastVisibleItemPosition: Int = layoutManager.findLastVisibleItemPosition()
        val totalItemCount = layoutManager.itemCount

        if (totalItemCount < previousTotalItemCount) {
            currentPage = startingPageIndex
            previousTotalItemCount = totalItemCount
            if (totalItemCount == 0) loading = true
        }

        if (loading && totalItemCount > previousTotalItemCount) {
            loading = false
            previousTotalItemCount = totalItemCount
        }

        if (!loading
            && lastVisibleItemPosition + visibleThreshold >= totalItemCount
            && dy > 0) {
            loading = if (currentPage < pages) {
                currentPage++
                onLoadMore(currentPage)
                true
            } else false
        }
    }

    abstract fun onLoadMore(page: Int)

    private fun resetPageCount(page: Int) {
        previousTotalItemCount = 0
        loading = true
        currentPage = page
        onLoadMore(currentPage)
    }

    fun resetPageCount() {
        resetPageCount(0)
    }

    fun setLoaded() {
        loading = false
    }

    fun setPages(pages: Int) {
        this.pages = pages
    }
}
